terraform/files/bin/fixup_k8sregistry.sh

```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+# fixup_k8s_version.sh
+# Patch $2 (cluster-template.yaml) with fixed up k8s imageRepo if needed
+# This is to find old kube-proxy, coredns, ... images on k8s.gcr.io and new ones on registry.k8s.io
+# See https://github.com/kubernetes-sigs/cluster-api/blob/main/internal/util/kubeadm/kubeadm.go#L40
+# (c) Kurt Garloff, Roman Hros, 02/2023
+# SPDX-License-Identifier: Apache-2.0
+
+if test -z "$2"; then echo "ERROR: Need clusterctl.yaml cluster-template args" 1>&2; exit 1; fi
+K8SVER=$(grep '^KUBERNETES_VERSION:' "$1" | sed 's/^KUBERNETES_VERSION: v\([0-9.]*\)/\1/')
+K8SMINOR=${K8SVER#*.}
+K8SPATCH=${K8SMINOR#*.}
+if test "$K8SPATCH" = "$K8SMINOR"; then K8SPATCH=0; fi
+K8SMINOR=${K8SMINOR%%.*}
+K8SVER=${K8SVER%%.*}$(printf %02i ${K8SMINOR})$(printf %02i ${K8SPATCH})
```

Comment:
As I found it hard to understand how the version is being read, what line 13 is doing and why, etc.:

Here is another option using yq and kind-of typed bash variables as much as possible (yq is available in this context, right?):

```
readarray -t -d "." K8SVER < <(yq eval '.KUBERNETES_VERSION | sub("v", "")' "$1" | tr -d '\n')
K8SMAJOR=${K8SVER[0]}
K8SMINOR=${K8SVER[1]}
K8SPATCH=${K8SVER[2]}
K8SVER=${K8SMAJOR}$(printf %02i ${K8SMINOR})$(printf %02i ${K8SPATCH})
```

Regarding actual comparison of versions: Of course, 1) reading the semver parts, 2) reformatting them to a special string and 3) using the string as integer to compare it against other special strings/integers, is pretty prone to errors.
So of course, using a semver library/binary would be optimal.
But all edge cases I can come up with, which could cause problems with the existing script are *probably* unlikely to happen in the near future: E. g. Kubernetes changing its release cadence to have fewer minor releases with many more patch releases, leading to versions like `1.30.101` which would cause problems here.

Anyways: Seems to work for this use case for now; If it should be really robust, IMHO considering to use a semver library like in the cluster-api code is the way to go. Alternatively, I imagine, such comparison could also be done without a library, when being done in Python: Just creating a simple semver class with attributes major/minor/patch and a simple `__lt__(self, other)` implementation attached. I'd be surprised if the same is possible with bash ;)

Reply:
> As I found it hard to understand how the version is being read, what line 13 is doing and why, etc.:

The bash shell substitution is not easy to read:
`${VAR%x}` returns `$VAR` with the shortest matching `x` removed from the end and the original `$VAR` if no `x` matches.
`${VAR#x}` does the same thing at the beginning, `%%` and `##` remove the longest matches at the end and beginning respectively.

If we are afraid of cumbersome code, I would turn to python instead rather than using another tool.
Yes, yq is available, but I'm not sure that everyone is familiar with it.

I would tend to keep it with hard-to-read bash for now, I'm not afraid of 3-digit patchlevels ...

Reply:
Line 13 specifically handles the case that there is no match, so we succeed parsing v1.24 as v1.24.0 => 12400. We would otherwise misparse to 12424.